### PR TITLE
LPD-37756 FDS is not applying user timezone when rendering dates

### PR DESCRIPTION
**Pull request by markocikos**
### References

- JIRA: [LPD-37756 FDS is not applying user timezone when rendering dates](https://issues.liferay.com/browse/LPD-37756)
- This is a followup on https://github.com/liferay-frontend/liferay-portal/pull/4458

### What is the goal of this PR?

We're adding API to set how FDS handles timezones: it can apply theme display time zone, or leave date as it is.

### What does it look like?

https://github.com/user-attachments/assets/ddabda7a-1354-4371-bfb5-63d6aa6ed545

image

### CODE REVIEW

File: modules/apps/frontend-data-set/frontend-data-set-api/src/main/java/com/liferay/frontend/data/set/constants/FDSTimeZoneBehaviors.java [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+/**
+ * @author Marko Cikos
+ */
+public class FDSTimeZoneBehaviors {
+
+	public static final String APPLY_THEME_DISPLAY_TIME_ZONE =
+		"applyThemeDisplayTimeZone";
+
+	public static final String DO_NOT_MODIFY_DATE_VALUE =
+		"doNotModifyDateValue";
```
Comment (markocikos) [inline modules/apps/frontend-data-set/frontend-data-set-api/src/main/java/com/liferay/frontend/data/set/constants/FDSTimeZoneBehaviors.java:17]: It might be a bit too verbose for typical naming patterns we have, but the idea to be as descriptive as possible.

File: modules/apps/frontend-data-set/frontend-data-set-api/src/main/java/com/liferay/frontend/data/set/view/table/BaseDateFDSTableSchemaField.java [hunk 2]
```diff
@@ -16,17 +17,31 @@ public JSONObject getFormat() {
...
 	@Override
 	public JSONObject toJSONObject() {
 		JSONObject jsonObject = super.toJSONObject();
 
-		return jsonObject.put("format", getFormat());
+		return jsonObject.put(
+			"format", getFormat()
+		).put(
+			"timeZoneBehavior", getTimeZoneBehavior()
+		);
```
Comment (markocikos) [inline modules/apps/frontend-data-set/frontend-data-set-api/src/main/java/com/liferay/frontend/data/set/view/table/BaseDateFDSTableSchemaField.java:40]: In difference to previous version, I set time zone behavior outside of `format`. The idea is that `format` structure matches second argument of `Intl.DateTimeFormat`.

File: modules/apps/frontend-data-set/frontend-data-set-sample-web/src/main/java/com/liferay/frontend/data/set/sample/web/internal/view/CustomizedTableFDSView.java [hunk 3]
```diff
@@ -155,6 +157,24 @@ public boolean isQuickActionsEnabled() {
...
+			"date"
+		).setLabel(
+			"date"
+		);
+
+		dateFDSTableSchemaField.setTimeZoneBehavior(
+			FDSTimeZoneBehaviors.APPLY_THEME_DISPLAY_TIME_ZONE);
+
+		return dateFDSTableSchemaField;
+	}
```
Comment (markocikos) [inline modules/apps/frontend-data-set/frontend-data-set-sample-web/src/main/java/com/liferay/frontend/data/set/sample/web/internal/view/CustomizedTableFDSView.java:176]: In difference to previous version, I set the usage example in our sample module, so we can safely push the solution. Commerce team can apply it to their module in a followup.

File: modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/cell_renderers/DateRenderer.tsx
Comment (markocikos) [inline modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/cell_renderers/DateRenderer.tsx:1]: I split the renaming to a separate commit, but still github separates them to delete + new. :disappointed:  I'm not sure if history will work.

**Comment by markocikos** [comment]
ci:test:relevant

**Comment by markocikos** [comment]
ci:stop

**Comment by markocikos** [comment]
ci:test:relevant

**Comment by markocikos** [comment]
ci:test:sf

**Comment by markocikos** [comment]
ci:test:relevant

**Comment by markocikos** [comment]
ci:test:relevant

File: modules/test/playwright/pages/users-admin-web/AccountSettingsPage.ts [hunk 4]
```diff
@@ -47,8 +49,28 @@ export class AccountSettingsPage {
 		]);
 	}
 
+	async goToDisplaySettings() {
+		await this.goToAccountSettings();
+
+		await this.page.locator('.nav-link', {hasText: 'Preferences'}).click();
```
Comment (smottal) [inline modules/test/playwright/pages/users-admin-web/AccountSettingsPage.ts:55]: can you create a locator for this like the this.userDisplayData one?

File: modules/test/playwright/pages/users-admin-web/AccountSettingsPage.ts [hunk 5]
```diff
@@ -47,8 +49,28 @@ export class AccountSettingsPage {
...
+			.click();
+	}
+
 	async selectAccountLanguage(option: string) {
 		await this.languageSelect.selectOption(option);
 		await this.saveButton.click();
 	}
+
+	async setTimeZone(timeZone: string) {
+		const timeZoneSelect = this.page.getByLabel('Time Zone');
```
Comment (smottal) [inline modules/test/playwright/pages/users-admin-web/AccountSettingsPage.ts:68]: also for this

File: modules/test/playwright/pages/users-admin-web/AccountSettingsPage.ts [hunk 6]
```diff
@@ -47,8 +49,28 @@ export class AccountSettingsPage {
...
+
+		await timeZoneSelect.selectOption(timeZone);
+
+		await this.page.getByRole('button', {name: 'Save'}).click();
```
Comment (smottal) [inline modules/test/playwright/pages/users-admin-web/AccountSettingsPage.ts:72]: this can be just `await this.saveButton.click()` as this locator already exists

**Comment by markocikos** [comment]
ci:test:sf

**Review by smottal — APPROVED** [review]

**Comment by smottal** [comment]
@markocikos perfect. Thx

File: modules/apps/frontend-data-set/frontend-data-set-sample-web/src/main/java/com/liferay/frontend/data/set/sample/web/internal/view/CustomizedTableFDSView.java (hunk 3 above)
Comment (dsanz) [reply modules/apps/frontend-data-set/frontend-data-set-sample-web/src/main/java/com/liferay/frontend/data/set/sample/web/internal/view/CustomizedTableFDSView.java:176]: this keeps the commerce tables just as they behave at this moment. We'd like to let commerce people know so that the make the adjustments later on.

For context for @smottal, origin of this issue is a customer who wanted to see the dates in the user-profile timezone. Rather than hardcoding the behavior in the FDS we're making it condfigurable by code. So now it's up to the different use cases to tell FDS how they want to render dates.

**Review by dsanz — APPROVED** [review]
LGTM ;) pls see comments inline

**Comment by markocikos** [comment]
ci:forward
